photosynthesis_metrics/vif.py

```diff
@@ -0,0 +1,138 @@
+    for scale in range(1, 5):
+        kernel_size = 2 ** (5 - scale) + 1
+        kernel = _gaussian_kernel2d(kernel_size, sigma=kernel_size / 5)
+        kernel = kernel.view(1, 1, kernel_size, kernel_size).to(prediction)
+
+        if scale > 1:
+            # Convolve and downsample
+            prediction = F.conv2d(prediction, kernel)[:, :, ::2, ::2]  # valid padding
+            target = F.conv2d(target, kernel)[:, :, ::2, ::2]  # valid padding
+
+        mu_trgt, mu_pred = F.conv2d(target, kernel), F.conv2d(prediction, kernel)  # valid padding
+        mu_trgt_sq, mu_pred_sq, mu_trgt_pred = mu_trgt * mu_trgt, mu_pred * mu_pred, mu_trgt * mu_pred
+
+        sigma_trgt_sq = F.conv2d(target ** 2, kernel) - mu_trgt_sq
+        sigma_pred_sq = F.conv2d(prediction ** 2, kernel) - mu_pred_sq
+        sigma_trgt_pred = F.conv2d(target * prediction, kernel) - mu_trgt_pred
+        
+        # Zero small negative values
+        sigma_trgt_sq[sigma_trgt_sq < 0] = 0
+        sigma_pred_sq[sigma_pred_sq < 0] = 0
+        
+        g = sigma_trgt_pred / (sigma_trgt_sq + EPS)
+        sigma_v_sq = sigma_pred_sq - g * sigma_trgt_pred
+
+        pred_vif_scale = torch.log10(1.0 + (g ** 2.) * sigma_trgt_sq / (sigma_v_sq + sigma_n_sq))
+        prediction_vif += torch.sum(pred_vif_scale, dim=[1, 2, 3])
+        target_vif += torch.sum(np.log10(1.0 + sigma_trgt_sq / sigma_n_sq), dim=[1, 2, 3])
+
+    return prediction_vif / target_vif
+
+
+class VIFLoss(_Loss):
+    r"""Creates a criterion that measures the Visual Information Fidelity loss
+    between predicted and target image. In order to be considered as a loss,
+    value `1 - clip(VIF, min=0, max=1)` is returned.
+    """
+
+    def __init__(self, sigma_n_sq: float = 2.0, data_range=1.0):
+        r"""
+        Args:
+            sigma_n_sq: HVS model parameter (variance of the visual noise).
+            data_range: Value range of input images (usually 1.0 or 255). Default: 1.0
+        """
+        super().__init__()
+        self.sigma_n_sq = sigma_n_sq
+        self.data_range = data_range
+
+    def forward(self, prediction: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
+        r"""Computation of Visual Information Fidelity (VIF) index as a loss function.
+        Args:
+            prediction: Tensor of prediction of the network.
+            target: Reference tensor.
+        Returns:
+            Value of VIF loss to be minimized. 0 <= VIFLoss <= 1.
+        """
+        prediction, target = _adjust_dimensions(x=prediction, y=target)
+        _validate_input(x=prediction, y=target)
+
+        return self.compute_metric(prediction, target)
+
+    def compute_metric(self, prediction: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
+
+        score = vif_p(prediction, target, sigma_n_sq=self.sigma_n_sq, data_range=self.data_range)
+        # Make sure value to be in [0, 1] range and convert to loss
+        loss = 1 - torch.clamp(torch.mean(score), 0, 1)
```

Comment:
In case of `SSIM` and `MS-SSIM`, we return the value, which should be maximised. What best practice should we use? Make all losses to be minimised or leave it for user?

Reply:
Yeah. I thin loss is intended to be minimised. We can either return `1 - metric` or write in the docs that user need to do that themselves.